File: Assign_ISO.R
## Assign ISO codes to DF
assign_iso <- function(df) {
	
	# Load ISO DF
	ISO <- read.csv("https://github.com/JeremyBowyer/Assign_ISO/raw/master/ISO%20Country%20Codes.csv")
	ISO[, grep("Name",names(ISO))] <- apply(ISO[, grep("Name",names(ISO))], 2, tolower)
	
	# create temporary data frame
	tempdf <- df
	columnorder <- c("Country", names(tempdf[-grep("Country", names(tempdf))]))
	tempdf <- tempdf[, columnorder]
	tempdf$Country <- tolower(tempdf$Country)
	tempdf$Country <- trimws(tempdf$Country)
	
	# merge tempdf with iso on multiple possible names
	suppressWarnings(tempdf_iso <- merge(tempdf, ISO[,c("Name.1", "Code")], by.x = "Country", by.y = "Name.1", all.x = TRUE))
	suppressWarnings(tempdf_iso <- merge(tempdf_iso, ISO[,c("Name.2", "Code")], by.x = "Country", by.y = "Name.2", all.x = TRUE))
	suppressWarnings(tempdf_iso <- merge(tempdf_iso, ISO[,c("Name.3", "Code")], by.x = "Country", by.y = "Name.3", all.x = TRUE))
	suppressWarnings(tempdf_iso <- merge(tempdf_iso, ISO[,c("Name.4", "Code")], by.x = "Country", by.y = "Name.4", all.x = TRUE))
	suppressWarnings(tempdf_iso <- merge(tempdf_iso, ISO[,c("Name.5", "Code")], by.x = "Country", by.y = "Name.5", all.x = TRUE))
	suppressWarnings(tempdf_iso <- merge(tempdf_iso, ISO[,c("Name.6", "Code")], by.x = "Country", by.y = "Name.6", all.x = TRUE))
	suppressWarnings(tempdf_iso <- merge(tempdf_iso, ISO[,c("Name.7", "Code")], by.x = "Country", by.y = "Name.7", all.x = TRUE))
	suppressWarnings(tempdf_iso <- merge(tempdf_iso, ISO[,c("Name.8", "Code")], by.x = "Country", by.y = "Name.8", all.x = TRUE))
	suppressWarnings(tempdf_iso <- merge(tempdf_iso, ISO[,c("Name.9", "Code")], by.x = "Country", by.y = "Name.9", all.x = TRUE))
	
	# rename columns
	isonames <- c("Name1", "Name2", "Name3", "Name4", "Name5", "Name6", "Name7", "Name8", "Name9")
	names(tempdf_iso) <- c(names(tempdf),isonames)
	
	# find first non-NA code
	tempdf_iso$ISO <- t(na.locf(t(tempdf_iso[,isonames]), na.rm = FALSE, fromLast = TRUE))[,1]
	
	# remove unnecessary iso code columns
	tempdf_iso <- tempdf_iso[,c(names(tempdf), "ISO")]
	
	# send final df
	tempdf_iso
}

File: Assign_ISO.py
import pandas as pd
import re
import numpy as np
isoCodes = pd.read_csv('https://raw.githubusercontent.com/JeremyBowyer/Assign_ISO/master/ISO%20Country%20Codes.csv')

def assign_iso(df, countryCol):

    # Find code for each possible country name in ISO code DF
    nameCols = [col for col in isoCodes.columns if re.compile("Name").match(col)]
    for col in nameCols:
        i = int(col.split("Name.")[1])
        df = df.merge(isoCodes[[col, "Code"]], how="left", left_on=countryCol, right_on=col, suffixes=["__"+str(i-1), "__"+str(i)])
        df.drop(col, axis=1, inplace=True)

    # Find first valid match, throw out all others
    codeCols = [col for col in df.columns if re.compile("Code__").match(col)]
    df['ISO'] = df[codeCols].apply(lambda row: next((val for val in list(row) if val == val), np.nan), axis=1)

    # Drop other code columns
    for col in codeCols:
        df.drop(col, axis=1, inplace=True)
        
    return df